src/presenter/utils.py

```diff
@@ -0,0 +1,110 @@
+from PyQt5 import QtCore
+from src import _IMAGES_STACK, IMAGES_STACK
+import numpy as np
+import copy
+# import sparse
+THREADING_ENABLED = False
+RUNNERS = []
+
+
+class Runner(QtCore.QThread):
+    """
+    QThread that activate a function with arguments
+
+    Parameters
+    ----------
+    target: function
+    *args, **kwargs: function arguments
+    """
+    def __init__(self, target, *args, **kwargs):
+        super().__init__()
+        self._target = target
+        self._args = args
+        self._kwargs = kwargs
+
+        # where the funciton result is stored
+        self.out = None
+
+    def run(self):
+        self.out = self._target(*self._args, **self._kwargs)
+
+
+def enable_threading(boolean):
+    global THREADING_ENABLED
+    THREADING_ENABLED = boolean
+
+
+def view_manager(threadable=True):
+    """
+    this decorator manage the loading gif and threading
+
+    Parameters
+    ----------
+    threadable: bool, default=True
+        if True, the model function will be processed inside a QThread (if allowed)
+
+    """
+    def decorator(foo):
+        def inner(presenter, widget):
+            # start gif animation
+            widget.node.gif.start()
+
+            function, args = foo(presenter, widget)
+
+            def updateView(output):
+                if isinstance(output, Exception):
+                    widget.node.gif.fail("[{0}] {1}".format(type(output).__name__, output))
+                else:
+                    store_image(output, widget.node.name)
+                    widget.node.updateSnap()
+                    widget.node.gif.stop()
+
+            # start the process inside a QThread
+            if threadable and THREADING_ENABLED:
+                runner = Runner(function, **args)
+                RUNNERS.append(runner)  # needed to keep a trace of the QThread
+                runner.finished.connect(lambda: updateView(runner.out))
+                runner.finished.connect(lambda: RUNNERS.remove(runner))
+                runner.start()
+            else:
+                updateView(function(**args))
+        return inner
+    return decorator
+
+
+def get_image(name):
+    if name not in _IMAGES_STACK:
+        print("'{}' not in image stack".format(name))
+        return None
+    # return _IMAGES_STACK[name].todense()
+    return _IMAGES_STACK[name]
+
+
+def store_image(im, name):
+    """
+    store raw sparsed image and (0, 255)-scaled image, 0 is nan values
+
+    Parameters
+    ----------
+    im: numpy.array
+    name: str
+    """
+    # initialize
+    # _IMAGES_STACK[name] = sparse.COO(im)
+    _IMAGES_STACK[name] = im
+    im_c = im.astype(np.float64) if im.dtype != np.float64 else copy.copy(im)
+    im_c[np.isinf(im_c)] = np.nan
+
+    # scale image in range (1, 255)
+    mini, maxi = np.nanmin(im_c), np.nanmax(im_c)
+    if mini == maxi:
+        mini = 0
+        maxi = max(maxi, 1)
+    im_c = (im_c - mini) / (maxi - mini) * 254 + 1
```

Comment:
You did not want to keep values as they come? For example with a mask image, 1 is transformed into 255.0, which is a bit confusing

Reply:
IMAGES_STACK is only for visualization.
To visualize the image in with the default colormap I need an 0-255 image. If I keep true values of binary images, we wont see the difference between 0 and 1 to the naked eyes..
So I need to scale images between 0 and 255 (0 represent infinite and nan values).

Reply:
I think that it is better to adapt colormap instead of changing the images. When I load label data and I click on a label to know its value, it says 255.0 instead of 1, that is not really good :/

Reply:
Ok its an error, it should show IMAGE_STACK but give the _IMAGE_STACK values when you click..
Can you give me an example when this issue append? I cant figure out how to get this issue

Reply:
I do not remember how I saw this behaviour, but now I cannot reproduce it ... We can say it is ok for now. If I encounter the issue again I will create an issue ;)